create_derivatives/routines.py

```diff
@@ -128,8 +132,121 @@ def ocr_pdf(self):
 
 
 class ManifestMaker:
-    # TO DO: make manifests
-    pass
+
+    def __init__(self):
+        self.server_url = settings.IMAGESERVER_URL
+        self.resource_url = "{}/iiif/3/".format(self.server_url)
+        self.fac = ManifestFactory()
+        self.fac.set_base_prezi_uri("{}/manifests/".format(self.server_url))
+        self.fac.set_base_image_uri(self.resource_url)
+        self.fac.set_debug(settings.PREZI_DEBUG)
+        self.upgrader = Upgrader()
+
+    def run(self):
+        bags_with_manifests = []
+        for bag in Bag.objects.filter(process_status=Bag.PDF):
+            jp2_files = sorted([str(f) for f in matching_files(str(Path(bag.bag_path, "data", "JP2")))])
+            manifest_dir = str(Path(bag.bag_path, "data", "MANIFEST"))
+            self.fac.set_base_prezi_dir(manifest_dir)
+            self.create_manifest(jp2_files, manifest_dir, str(Path(bag.bag_path, "data", "JP2")), bag.dimes_identifier, bag.as_data)
+            bag.process_status = Bag.MANIFESTS_CREATED
+            bag.save()
+            bags_with_manifests.append(bag.dimes_identifier)
+        msg = "Manifests successfully created." if len(bags_with_manifests) else "No manifests created."
+        return msg, bags_with_manifests
+
+    def create_manifest(self, files, manifest_dir, image_dir, identifier,
+                        obj_data):
+        """Method that runs the other methods to build a manifest file and populate
+        it with information.
+
+        Args:
+            files (list): Files to iterate over
+            manifest_dir (str): Path to directory to write manifests to.
+            image_dir (str): Path to directory containing derivative image files.
+            identifier (str): A unique identifier.
+            obj_data (dict): Data about the archival object.
+        """
+        manifest_path = "{}.json".format(str(Path(manifest_dir, identifier)))
+        page_number = 1
+        manifest = self.fac.manifest(ident="{}{}".format(self.fac.prezi_base, identifier), label=obj_data["title"])
+        manifest.set_metadata({"Date": obj_data["dates"]})
+        manifest.thumbnail = self.set_thumbnail(Path(files[0]).stem)
+        sequence = manifest.sequence(ident=identifier)
+        for file in files:
+            page_ref = Path(file).stem
+            width, height = self.get_image_info(image_dir, file)
+            canvas = sequence.canvas(
+                ident="{}/canvas/{}".format(
+                    manifest.id, str("{0:03}".format(page_number))),
```

Comment:
We're setting the canvas id, which starts the same as the manifest id, and then includes the canvas "page number" which iterates up by one for each file. The `{0:03}` is coercing it to three digits, and then it's being transformed into a string for inclusion into the id.

Reply:
OK, so a couple of followups then:
- Why are we adding leading zeroes? (I know this is related to #31, just trying to understand current thinking)
- If we're just adding leading zeros, why aren't we just using the page number from the original file?
- If there is a good reason for having two different methods of creating page numbers, can we use `zfill` instead of the list slicing?

Reply:
> OK, so a couple of followups then:
> 
> * Why are we adding leading zeroes? (I know this is related to [Review approach to page number parsing/creation across app #31](https://github.com/RockefellerArchiveCenter/pictor/issues/31), just trying to understand current thinking)
> * If we're just adding leading zeros, why aren't we just using the page number from the original file?
> * If there is a good reason for having two different methods of creating page numbers, can we use `zfill` instead of the list slicing?

- Used leading zeroes because I thought we wanted it to be in line with the the page numbers from the files. However, if we don't want them, we'd want to change the information that the files are giving us.
- Wasn't just using the page number from the files because originally we had the option to skip the first page of a file because it was just a "target" from reference, or in case there were inconsistencies or jumps in the file numbering (001, 002, 004). That first point isn't really necessary anymore, but the second one might be? I'm not sure if we care about that though.
- Didn't know about zfill, but happy to implement it if we want to go with leading zeroes.

Reply:
OK, for now can you throw in an inline comment at that line explaining what this is doing and why so that's documented?